modules/xmpp/Lobby.js

```diff
@@ -0,0 +1,320 @@
+                (roomJid, from, txt, invitePassword) => {
+                    logger.debug(`Received approval to join ${roomJid} ${from} ${txt}`);
+                    if (roomJid === this.mainRoom.roomjid) {
+                        // we are now allowed let's join and leave lobby
+                        this.mainRoom.join(invitePassword);
+
+                        this._leaveLobbyRoom();
+                    }
+                });
+            this.lobbyRoom.addEventListener(
+                XMPPEvents.MUC_DESTROYED,
+                (reason, jid) => {
+                    // we are receiving the jid of the main room
+                    // means we are invited to join, maybe lobby was disabled
+                    if (jid && jid === this.mainRoom.roomjid) {
+                        this.mainRoom.join();
+
+                        return;
+                    }
+
+                    this.mainRoom.eventEmitter.emit(XMPPEvents.MUC_DESTROYED, reason);
+                });
+
+            // If participant retries joining shared password while waiting in the lobby
+            // and succeeds make sure we leave lobby
+            this.mainRoom.addEventListener(
+                XMPPEvents.MUC_JOINED,
+                () => {
+                    this._leaveLobbyRoom();
+                });
+        }
+
+        return new Promise((resolve, reject) => {
+            this.lobbyRoom.addEventListener(XMPPEvents.MUC_JOINED, () => {
+                resolve();
+
+                // send our email, as we do not handle this on initial presence we need a second one
+                if (email && !isModerator) {
+                    this.lobbyRoom.removeFromPresence(EMAIL_COMMAND);
+                    this.lobbyRoom.addToPresence(EMAIL_COMMAND, { value: email });
+                    this.lobbyRoom.sendPresence();
+                }
+            });
+            this.lobbyRoom.addEventListener(XMPPEvents.ROOM_JOIN_ERROR, reject);
+            this.lobbyRoom.addEventListener(XMPPEvents.ROOM_CONNECT_NOT_ALLOWED_ERROR, reject);
+            this.lobbyRoom.addEventListener(XMPPEvents.ROOM_CONNECT_ERROR, reject);
+
+            this.lobbyRoom.join();
+        });
+
+    }
+
+    /**
+     * Should be possible only for moderators.
+     * @param id
+     */
+    denyAccess(id) {
+        if (!this.isSupported() || !this.mainRoom.isModerator()) {
+            return;
+        }
+
+        const jid = Object.keys(this.lobbyRoom.members)
+            .find(j => Strophe.getResourceFromJid(j) === id);
+
+        if (jid) {
+            this.lobbyRoom.kick(jid);
+        } else {
+            logger.error(`Not found member for ${id} in lobby room.`);
+        }
+    }
+
+    /**
+     * Should be possible only for moderators.
+     * @param id
+     */
+    approveAccess(id) {
+        if (!this.isSupported() || !this.mainRoom.isModerator()) {
+            return;
+        }
+
+        const roomJid = Object.keys(this.lobbyRoom.members)
```

Comment:
That's the JID of the member to approve, not the JID of the room, right?

Reply:
Yep.